File: 438_FindAllAnagraInAString.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        
        List<Integer> result = new ArrayList<>();
		if(s.length() == 0 || p.length() == 0)
			return result;

		// maintain a table whose size is 256,
		int[] table = new int[256]; 

		for(int i = 0; i < p.length(); ++i){
			table[p.charAt(i)]++;
		}

		int left = 0, right = 0;
		int cnt = p.length();

		while(right < s.length()){
			if(table[s.charAt(right)] >= 1)
				--cnt;

			table[s.charAt(right)]--;
			right++;

			if(cnt == 0)
				result.add(left);

			if(right - left == p.length()){
				if(table[s.charAt(left)] >= 0)
					++cnt;

				table[s.charAt(left)]++;
				left++;
			}
		}
        
        return result;
        
    }
}

File: 438_FindAllAnagraInAString.cc
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        if(s.size() == 0)
            return {};
        
        vector<int> res;
        vector<int> m(256, 0);
        
        int left = 0, right = 0;
        int cnt = p.size();
        int n = s.size();
        
        for(auto& c: p)
            ++m[c];
        
        while(right < n){
            if(m[s[right++]]-- >= 1)
                --cnt;
            if(cnt == 0)
                res.push_back(left);
            
            if(right - left == p.size() && m[s[left++]]++ >= 0)
                ++cnt;
        }
        return res;
    }
};
